query.go

```diff
@@ -396,15 +402,42 @@ func (q *query) queryPeer(ctx context.Context, ch chan<- *queryUpdate, p peer.ID
 		ch <- &queryUpdate{cause: p, unreachable: []peer.ID{p}}
 		return
 	}
-
 	queryDuration := time.Since(startQuery)
+	// return if we got a nil response
+	// This happens if the self peer decides to not run the query RPC on the remote peer for some reason
+	if newPeers == nil {
+		ch <- &queryUpdate{cause: p, heard: nil, queried: []peer.ID{p}, queryDuration: queryDuration}
+		return
+	}
 
 	// query successful, try to add to RT
 	q.dht.peerFound(q.dht.ctx, p, true)
 
 	// process new peers
-	saw := []peer.ID{}
-	for _, next := range newPeers {
+	saw := make(map[peer.ID]struct{})
+
+	// process signed peer records
+	for pid, record := range newPeers.signedPeerRecords {
+		if pid == q.dht.self { // don't add self.
+			logger.Debugf("PEERS CLOSER -- worker for: %v found self", p)
+			continue
+		}
+
+		// add any other know addresses for the candidate peer.
+		curInfo := q.dht.peerstore.PeerInfo(pid)
```

Comment:
Why are we doing this? If the signed record on its own doesn't warrant being queried on its own then why should we help it by adding externally learned addresses?